### DISCUSSION TITLE
Handling submit errors when using React Query.

### DISCUSSION THREAD

**Discussion by adriantrunzo** [react-hook-form/react-hook-form#6574] [Q&A]
Hi,

I have a project that uses both React Hook Form and React Query (two amazing projects!). We have settled on a pattern for calling a mutation from within `handleSubmit` such that React Hook Form can be aware of the state of the asynchronous operation, but management of the promise stays in the mutation definition. Here's a codesandbox example: https://codesandbox.io/s/rhf-rq-submit-async-2jjmf?file=/src/App.js

Our use of `handleSubmit` is not very intuitive, but necessary due to the fact that [handleSubmit will rethrow errors](https://github.com/react-hook-form/react-hook-form/blob/12561a713941f34545131af3dc437aae19bea641/src/logic/createFormControl.ts#L1051):

```jsx
const handleFormSubmit = async (event) => {
  try {
    await handleSubmit(mutateAsync)(event);
  } catch {
    // Do nothing. Handled by mutation.
  }
}
```

Is there a simpler way to handle this situation? I am tempted to request a setting in `useForm` that causes handleSubmit to not re-throw errors, but I don't know how feasible or widely useful that would be. For example:

```jsx
const { handleSubmit } = useForm({ handleSubmitThrowError: false }) // true by default
```

And then in the definition of `handleSubmit`, you'd have:

```jsx
} catch (err) {
  hasNoPromiseError = false;

  if (handleSubmitThrowError) {
    throw err;
  }
}
```

Then we could just use `handleSubmit` in our form as per usual:

```jsx
<form onSubmit={handleSubmit(mutateAsync)}>
```

Thanks!

**Comment by bluebill1049** [comment]
I think we had a similar discussion on this one, yet we have come up with a good solution for this.
